pkg/pmem-csi-driver/controllerserver.go

```diff
@@ -161,20 +167,44 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		volumeID := id.String()
 		if cs.mode == Unified || cs.mode == Node {
 			glog.Infof("CreateVolume: Special create volume in Unified mode")
-			if err := cs.dm.CreateDevice(volumeID, asked, nsmode); err != nil {
+			if err := cs.dm.CreateDevice(volumeID, uint64(asked), nsmode); err != nil {
 				return nil, status.Errorf(codes.Internal, "CreateVolume: failed to create volume: %s", err.Error())
 			}
-		} else /*if cs.mode == Unified */ {
-			foundNode := false
-			for _, nodeInfo := range cs.rs.nodeClients {
-				if nodeInfo.Capacity[nsmode] >= asked {
-					foundNode = true
-					break
+			topology = append(topology, &csi.Topology{
+				Segments: map[string]string{
+					"kubernetes.io/hostname": cs.Driver.nodeID,
+				},
```

Comment:
This is where it becomes interesting. A CSI driver shouldn't need to know about Kubernetes host names. It's the other way around: the CSI driver has its own local host name, and the sidecar containers take care of mapping that to the Kubernetes host name.

According to Michelle, external-provisioner does that based on the information in `CSIDriverInfo`. Without `CSIDriverInfo` there should be no `NodeAffinity` in the `PV`.

@avalluri Were you testing with the `CSINodeInfo` CRD installed and the corresponding feature gate enabled?

My current theory is that `external-provisioner` just passed through the CSI driver values. Whether it should do that is open for debate...

Reply:
The change tested with external-provisioner v4.0.1, I don't think that CSINodeInfo is not implemented in this version.